ILL Reflectometry workflow

gvardany opened the pull request:
**Description of work.**

This is the first cut implementation of the high level steering algorithm for TOF specular reflectometry reduction for the ILL instruments D17 and Figaro. 

Features incoherent and coherent reductions.

While polarisation is foreseen, it is not yet tested; there will be a potential second PR to wrap up that aspect.

Some details in the calculations might need further polishing, but we need to release this to the scientists so that they can give a spin.

**To test:**
Skim through the documentation, play with the usage examples. Particularly run with incoherent and coherent modes, give a Bragg angle by value or alternate between Sample Angle and Detector Angle options.

<!-- Instructions for testing. -->

Fixes #24555 . <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

gvardany commented:
Seems like Stitch fails on linux, undoing the last check in the doc test, until proper investigation why.

mantid-builder commented:
@gvardany have you had a chance to see why the build is failing?
<!-- {"problem_type": "failing"} -->

mantid-builder commented:
@gvardany the build is failing, can you investigate.
<!-- {"problem_type": "failing"} -->

gvardany commented:
@gemmaguest This is ready for another look.

gvardany commented on docs/source/release/v4.1.0/reflectometry.rst:
```diff
@@ -2,6 +2,9 @@
 Reflectometry Changes
```
I don't really think the copyright date is significant, regarding the test files, yes you need a lot to do a meaningful reduciton, what I had to do is to copy the same files from system test data also to unit and doc test data, since apparently the doc tests and usage examples are run without system test data.

gemmaguest replied:
Are all of the test files needed in each of those locations? It looks like only a small subset are used in the unit tests? And it looks like some (e.g. 397805.nxs) are not used at all? Unless I'm missing something!

gemmaguest replied:
Regarding the release notes, it looks like the content has been moved to v4.3 but there are some formatting changes remaining in v4.1. Could you revert those just to keep things tidy? I find it confusing to see changes to old release notes!

gvardany replied:
I think the copy of files in docs tests is probably not needed, I will try to remove them. It's true, not all of the files are used in the current tests, I thought about writing more tests later, when it's more stabilized, but prepare the data already, but maybe I should indeed remove those for the moment. As for the release notes, there was some title level mismatch in 4.1 that I had to fix to get the build passing without warnings.

gemmaguest commented:
I don't think it's related to this PR, but just to note, if I unroll `ExtractMonitors` in the history then I can't re-run the script. I think that `ExtractMonitors` is not setting the correct output name when it calls `ExtractSpectra` - e.g. I'd expect `317369_direct_detectors_` here not `detector_ws`:
```
ExtractSpectra(InputWorkspace='317369_direct_merged_', OutputWorkspace='detector_ws', WorkspaceIndexList='0-255')
ExtractSpectra(InputWorkspace='317369_direct_merged_', OutputWorkspace='monitor_ws', WorkspaceIndexList='256-257')
```

gemmaguest commented:
I also can't replay the history if I just unroll one of the direct child algorithms e.g. `ReflectometryILLPreprocess`. Would you expect this to work?

gemmaguest commented:
When I run `ReflectometryILLAutoProcess` from the algorithm dialog, double values have strange defaults set:

![Capture](https://user-images.githubusercontent.com/12895056/72524097-13950700-3859-11ea-8352-dc4d613fdc74.PNG)

These get used in the reduction giving me unexpected results.

gemmaguest commented:
Is it worth adding the new algorithm to the "see also" list for its child algorithms?

gvardany commented:
Regarding the strange default values of the properties, this is because they are all array properties, hence `isDefault` can not be checked on those, I think. Meaning that they have to have some initial value in order to check whether they are set, or left to default. I agree though that the maxdbl is not the best for this. I don't know why it was chosen in the beginning. I'll try to set more meaningful defaults, like 0 or negative, and see how that goes.

gvardany commented:
I did not think about unrolling the history, I will have a look into this.

gemmaguest commented:
> I did not think about unrolling the history, I will have a look into this.

In case it causes confusion: note that there's a bug unrolling algorithm history for two consecutive workflow algorithms (#27492). But I think if you just have one top level workflow algorithm and unroll its children then it should not be affected by that bug.

gvardany commented:
> I don't think it's related to this PR, but just to note, if I unroll `ExtractMonitors` in the history then I can't re-run the script. I think that `ExtractMonitors` is not setting the correct output name when it calls `ExtractSpectra` - e.g. I'd expect `317369_direct_detectors_` here not `detector_ws`:
> 
> ```
> ExtractSpectra(InputWorkspace='317369_direct_merged_', OutputWorkspace='detector_ws', WorkspaceIndexList='0-255')
> ExtractSpectra(InputWorkspace='317369_direct_merged_', OutputWorkspace='monitor_ws', WorkspaceIndexList='256-257')
> ```

Can you point out to the line where this happens?

gemmaguest commented:
> > I don't think it's related to this PR, but just to note, if I unroll `ExtractMonitors` in the history then I can't re-run the script. I think that `ExtractMonitors` is not setting the correct output name when it calls `ExtractSpectra` - e.g. I'd expect `317369_direct_detectors_` here not `detector_ws`:
> > ```
> > ExtractSpectra(InputWorkspace='317369_direct_merged_', OutputWorkspace='detector_ws', WorkspaceIndexList='0-255')
> > ExtractSpectra(InputWorkspace='317369_direct_merged_', OutputWorkspace='monitor_ws', WorkspaceIndexList='256-257')
> > ```
> 
> Can you point out to the line where this happens?

I'm not quite sure how the names get set but I would guess [here](https://github.com/gemmaguest/mantid/blob/master/Framework/PythonInterface/plugins/algorithms/ExtractMonitors.py#L77) and [here](https://github.com/gemmaguest/mantid/blob/master/Framework/PythonInterface/plugins/algorithms/ExtractMonitors.py#L84)

gvardany commented:
I think I addressed your comments. Please have another look, especially make sure that the strange default values are not picked up accidentally. I've changed the defaults to -1, and invoke `isDefault` to check whether is user given or left to default.

gemmaguest commented:
The default values problem is fixed, as is the unrolling of `ExtractMonitors` :+1:

However, I still cannot replay the history if I unroll `ReflectometryILLPreprocess`, e.g.
- Run the first usage example:
```
config['default.facility'] = 'ILL'
config['default.instrument'] = 'D17'
config.appendDataSearchSubDir('ILL/D17/')

ws = ReflectometryILLAutoProcess(
 Run='317370',
 DirectRun='317369',
 WavelengthLowerBound=3.5,
 WavelengthUpperBound=24.5,
 DeltaQFractionBinning=0.1
)
```
- Show history for `ws_0` and tick Unroll for `ReflectometryILLPreprocess`
- Copy the history to clipboard, paste it into a script window and run
- I get this script, which fails because the `317369_direct` input to `ReflectometryILLSumForeground` does not exist.
```

# Child algorithms of ReflectometryILLAutoProcess

## Child algorithms of ReflectometryILLPreprocess
LoadAndMerge(Filename='/home/gemma/work/checkout/mantid/build/ExternalData/Testing/Data/UnitTest/ILL/D17/317369.nxs', LoaderName='LoadILLReflectometry', LoaderOptions='{"BeamCentre":127.5,"BraggAngle":0,"XUnit":"TimeOfFlight"}', OutputWorkspace='__317369_direct_merged_')
ExtractMonitors(InputWorkspace='__317369_direct_merged_', DetectorWorkspace='__317369_direct_detectors_', MonitorWorkspace='__317369_direct_monitors_')
DeleteWorkspace(Workspace='__317369_direct_merged_')
FindReflectometryLines(InputWorkspace='__317369_direct_detectors_', OutputWorkspace='__317369_direct_peak_', EndWorkspaceIndex=255)
DeleteWorkspace(Workspace='__317369_direct_peak_')
SpecularReflectionPositionCorrect(InputWorkspace='__317369_direct_detectors_', TwoTheta=1.5885926485061646, DetectorCorrectionType='RotateAroundSample', DetectorComponentName='detector', OutputWorkspace='__317369_direct_detectors_moved_', DetectorFacesSample=True, PixelSize=0.0011950000000000001)
DeleteWorkspace(Workspace='__317369_direct_detectors_')
Scale(InputWorkspace='__317369_direct_detectors_moved_', OutputWorkspace='__317369_direct_normalised_to_slits_', Factor=6.5504725211368902)
DeleteWorkspace(Workspace='__317369_direct_detectors_moved_')
Scale(InputWorkspace='__317369_direct_normalised_to_slits_', OutputWorkspace='__317369_direct_normalised_to_time_', Factor=0.0016664481555508925)
DeleteWorkspace(Workspace='__317369_direct_normalised_to_slits_')
DeleteWorkspace(Workspace='__317369_direct_monitors_')
CloneWorkspace(InputWorkspace='__317369_direct_normalised_to_time_', OutputWorkspace='__317369_direct_cloned_for_flat_bkg_')
Transpose(InputWorkspace='__317369_direct_cloned_for_flat_bkg_', OutputWorkspace='__317369_direct_transposed_clone_')
DeleteWorkspace(Workspace='__317369_direct_cloned_for_flat_bkg_')
CalculatePolynomialBackground(InputWorkspace='__317369_direct_transposed_clone_', OutputWorkspace='__317369_direct_transposed_flat_background_', XRanges='189.5,194.5,209.5,214.5', CostFunction='Unweighted least squares')
DeleteWorkspace(Workspace='__317369_direct_transposed_clone_')
Transpose(InputWorkspace='__317369_direct_transposed_flat_background_', OutputWorkspace='__317369_direct_flat_background_')
DeleteWorkspace(Workspace='__317369_direct_transposed_flat_background_')
Minus(LHSWorkspace='__317369_direct_normalised_to_time_', RHSWorkspace='__317369_direct_flat_background_', OutputWorkspace='__317369_direct_flat_background_subtracted_')
DeleteWorkspace(Workspace='__317369_direct_normalised_to_time_')
DeleteWorkspace(Workspace='__317369_direct_flat_background_')
ConvertUnits(InputWorkspace='__317369_direct_flat_background_subtracted_', OutputWorkspace='__317369_direct_in_wavelength_', Target='Wavelength')
DeleteWorkspace(Workspace='__317369_direct_flat_background_subtracted_')
DeleteWorkspace(Workspace='__317369_direct_in_wavelength_')
## End of child algorithms of ReflectometryILLPreprocess

ReflectometryILLSumForeground(InputWorkspace='317369_direct', OutputWorkspace='317369_direct_frg', WavelengthRange='3.5,24.5')
ReflectometryILLPreprocess(Run='/home/gemma/work/checkout/mantid/build/ExternalData/Testing/Data/UnitTest/ILL/D17/317370.nxs', OutputWorkspace='317370_reflected', DirectLineWorkspace='317369_direct', ForegroundHalfWidth='0,0')
ReflectometryILLSumForeground(InputWorkspace='317370_reflected', OutputWorkspace='317370_reflected_frg', DirectForegroundWorkspace='317369_direct_frg', DirectLineWorkspace='317369_direct', WavelengthRange='3.5,24.5')
ReflectometryILLConvertToQ(InputWorkspace='317370_reflected_frg', OutputWorkspace='ws_0', DirectForegroundWorkspace='317369_direct_frg', GroupingQFraction=0.10000000000000001)
GroupWorkspaces(InputWorkspaces='ws_0', OutputWorkspace='ws')
DeleteWorkspace(Workspace='317369_direct_frg')
DeleteWorkspace(Workspace='317370_reflected_frg')
DeleteWorkspace(Workspace='317370_reflected')
DeleteWorkspace(Workspace='317369_direct')
# End of child algorithms of ReflectometryILLAutoProcess
```

gvardany commented:
Given that this PR is a year old, I would suggest to fix that in a separate issue. Anyways, here no one is going to replay the unrolled history.

gvardany commented:
For the reference, it's here: https://github.com/mantidproject/mantid/issues/27808
